node/src/components/consensus/protocols/highway.rs

```diff
@@ -404,6 +412,43 @@ impl<I: NodeIdT, C: Context + 'static> HighwayProtocol<I, C> {
         }
         dropped_vertices
     }
+
+    /// Returns an iterator over all the values that are expected to become finalized, but are not
+    /// finalized yet.
+    pub(crate) fn non_finalized_values<'a>(
+        &'a self,
+        fork_choice: Option<C::Hash>,
+    ) -> Box<dyn Iterator<Item = &'a C::ConsensusValue> + 'a> {
+        // early return if there is no fork choice
+        let fork_choice = match fork_choice {
+            Some(bhash) => bhash,
+            None => {
+                return Box::new(iter::empty());
+            }
+        };
+        let next_finalized_height = self
+            .finality_detector
+            .last_finalized()
+            .map_or(0, |bhash| self.highway.state().block(bhash).height + 1);
+        Box::new(
+            // start at the next finalized height
+            (next_finalized_height..)
+                // take the ancestor of the fork choice at this height
+                .map(move |height| {
+                    self.highway
+                        .state()
+                        .find_ancestor(&fork_choice, height)
+                        .cloned()
+                })
+                // only take values while there actually exists such an ancestor
+                .take_while(Option::is_some)
+                // flatten the `Option` returned by the iterator
+                .flatten()
+                // the ancestor is expressed as a block hash - return the value contained in the
+                // block
+                .map(move |bhash| &self.highway.state().block(&bhash).value),
```

Comment:
Wouldn't it be easier to start from the fork choice and iterate until the `last_finalized` using the unit's parents?

Reply:
More efficient, at least :+1: It would only be a problem if we ever started caring about the order of the values returned here, but that doesn't look likely.

Reply:
Yes, I meant "more efficient" :) We don't need to call `find_ancestor` for every intermediate unit.

Reply:
It also seems simpler to understand to me.